context/handlers.go

```diff
@@ -16,3 +34,12 @@ func TenantIDHandler(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r.WithContext(ContextWithTenantID(r.Context(), tenantID)))
 	})
 }
+
+func MaxConnectionAgeHandler(next http.Handler, maxConnectionAge time.Duration) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if start, ok := ConnectionStartTimeFromContext(r.Context()); ok && clock.Now().After(start.Add(maxConnectionAge)) {
+			w.Header().Add("Connection", "close")
```

Comment:
This relies on the http client respecting the `Connection: close` header, right? Do we know for sure that it respects that?

Reply:
I took a look at the golang server code to verify, but this should work.

https://github.com/golang/go/blob/d7ab277eed4d2e5ede4f3361adf42d4ad76ced8f/src/net/http/server.go#L452-L456

Reply:
It tried it e2e with Prometheus and it behaved exactly as expected.

The `ConnState` hook was super handy to understand what's going on:

```diff
+               ConnState: func(c net.Conn, s http.ConnState) {
+                       _ = level.Debug(logger).Log("msg", "connection state changed", "remote-addr", c.RemoteAddr(), "state", s)
+               },
```